### PR TITLE
PLUGIN-119: Partitioning support for BigQuery Sink (update operation on existing table)

### PR DESCRIPTION
**Pull request by nbesimi**

### CODE REVIEW

File: examples/example-bigquery-to-bigquery-update-partition-filter-enabled-cdap-data-pipeline.json [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                        "truncateTable": "false",
+                        "allowSchemaRelaxation": "false",
+                        "location": "US",
+                        "createPartitionedTable": "false",
+                        "partitionFilterRequired": "false",
+                        "schema": "${inputSchema}",
+                        "referenceName": "BigQuerySinkExample",
+                        "table": "${sinkTable}",
+                        "relationTableKey": "${tableKey}",
+                        "partitionFilter": "${partitionFilter}",
```
Comment (CuriousVini) [inline examples/example-bigquery-to-bigquery-update-partition-filter-enabled-cdap-data-pipeline.json:81]: how does the filter look like in this case?

File: docs/BigQueryTable-batchsink.md [hunk 2]
```diff
@@ -55,6 +55,9 @@ Should only be used with the Insert operation.
 multiple input records with the same key. For example, if this is set to 'updated_time desc', then if there are
 multiple input records with the same key, the one with the largest value for 'updated_time' will be applied.
 
+**Partition Filter**: Partition filter that can be used for partition elimination during Update operation.
+Should only be used with Update operation for tables where require partition filter is enabled.
```
Comment (CuriousVini) [inline docs/BigQueryTable-batchsink.md]: please also add example of what partition filter value should be since its a text box.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java [hunk 3]
```diff
@@ -164,6 +173,21 @@ public String getDedupeBy() {
...
 
+  @Nullable
+  public String getPartitionFilter() {
+    if (partitionFilter != null) {
+      partitionFilter = partitionFilter.trim();
+      if (partitionFilter.isEmpty()) {
+        return null;
+      }
+      // remove the WHERE from the filter
+      if (partitionFilter.toUpperCase().startsWith(WHERE)) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java]: why is this needed?

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java [hunk 4]
```diff
@@ -164,6 +173,21 @@ public String getDedupeBy() {
     return Strings.isNullOrEmpty(dedupeBy) ? null : dedupeBy;
   }
 
+  @Nullable
+  public String getPartitionFilter() {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java:177]: does this need to be public?

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java [hunk 5]
```diff
@@ -164,6 +173,21 @@ public String getDedupeBy() {
...
+      partitionFilter = partitionFilter.trim();
+      if (partitionFilter.isEmpty()) {
+        return null;
+      }
+      // remove the WHERE from the filter
+      if (partitionFilter.toUpperCase().startsWith(WHERE)) {
+        partitionFilter = partitionFilter.replace(WHERE, "");
+      }
+    }
+    return partitionFilter;
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java]: better have it reverse
```
if (Strings.isNullOrEmpty(partitionFilter)) {
 return partitionFilter;
}
...

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java [hunk 6]
```diff
@@ -496,5 +506,18 @@ protected void cleanup(JobContext context) throws IOException {
           .execute();
       }
     }
+
+    private static String formatPartitionFilter(String partitionFilter) {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:508]: please provide what is the input and output this method in the javadoc

File: examples/example-bigquery-to-bigquery-update-partition-filter-enabled-cdap-data-pipeline.json (hunk 1 above)
Comment (nbesimi) [reply examples/example-bigquery-to-bigquery-update-partition-filter-enabled-cdap-data-pipeline.json:81]: This represents the partition elimination for the update/upsert operations on existing tables where partition filter is enabled. Examples:
1. If you have partitioned table by time:
`_PARTITIONTIME > "2020-01-01"`
`_PARTITIONTIME > "2020-01-01" AND _PARTITIONTIME < "2020-03-01"`

2. If you have partitioned table partitioned by Integer value (e.x. gameId):
`gameId < 5`
`gameId > 1 AND gameId < 7`

File: docs/BigQueryTable-batchsink.md (hunk 2 above)
Comment (nbesimi) [reply docs/BigQueryTable-batchsink.md]: @CuriousVini we are going to add the following example:

> For example, if the table is partitioned the Partition Filter '_PARTITIONTIME > "2020-01-01" and _PARTITIONTIME < "2020-03-01"', the update operation will be performed in the partitions meeting the criteria.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java (hunk 3 above)
Comment (nbesimi) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java]: @CuriousVini, this is needed in case someone copy/paste or accidentally puts the WHERE keyword at the beginning of the Filter Partition textbox.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java (hunk 4 above)
Comment (nbesimi) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java:177]: @CuriousVini yes this needs to be public since it is used in `BigQuerySink.java`

```
if (config.getPartitionFilter() != null) {
      baseConfiguration.set(BigQueryConstants.CONFIG_PARTITION_FILTER, getConfig().getPartitionFilter());
    }
```

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java (hunk 5 above)
Comment (nbesimi) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQuerySinkConfig.java]: @CuriousVini will refactor the above to:

```
if (Strings.isNullOrEmpty(partitionFilter)) {
  return null;
}
partitionFilter = partitionFilter.trim();
// remove the WHERE keyword from the filter if the user adds it at the begging of the expression
if (partitionFilter.toUpperCase().startsWith(WHERE)) {
   partitionFilter = partitionFilter.substring(WHERE.length());
}
return  partitionFilter;
```

File: src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java (hunk 6 above)
Comment (nbesimi) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:508]: @CuriousVini can you please provide us some guidelines on how to do this

Comment (CuriousVini) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java:508]: Since this is modifying the filter configuration, could you add what is the user provided input and how this method modifies it and what is the resultant output

**Review by CuriousVini — APPROVED** [review]
just one comment about javadoc. LGTM
